### ISSUE TITLE
TS2536 only when type argument is a generic type parameter

### ISSUE THREAD

**Issue by ryan-0324** [microsoft/TypeScript#59000]
### 🔎 Search Terms

ts2536, ts2536 generic, generic constraint no effect

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generic constraints.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play/?#code/PTAEBUE8AcFNQOQCIBmB7NSGgMYEMA7AtAF1ACN4BXAZ1gBNQS1QBLA+2ADyZngQAKeAE4lWeADYB5AgFUOsFOwYAecAD4EAOlAAKAEwBWAMwA2AJQAoEn1ABxWAVjDWONaG4lH9GqADeoOhoAFygNCQuBADmANygAL7qoAC8oEKi4tJyCkpO9GrqANqoGEgAujGWliCgAHIszsJowta29QQOTi44KWkiYpIy8py5qgFBoeGRsQlFJZgVVTZwfRmD2SPK+RqWoHu9AYUA0rCQbASgANanaCgQoHi+VDlbHlxeHL7gx6dloAD8oCcADdnKBQidIGVQt9IRUErt9gAyfygH5ndhXG53cAPJ4vPJvD4+CDov6AyHgoGwUHCMr-GFkuLxSpAA)

### 💻 Code

```ts
// Type '"foo"' cannot be used to index type 'PartialOnUndefined<T>'. (2536)
type Generic<T extends { foo: string; }> = PartialOnUndefined<T>["foo"];

// No error
type NonGeneric = PartialOnUndefined<{ foo: string; }>["foo"];

type PartialOnUndefined<T>
    = { [Key in keyof T as undefined extends T[Key] ? never : Key]: T[Key]; }
    & { [Key in keyof T as undefined extends T[Key] ? Key : never]?: T[Key]; };
```

### 🙁 Actual behavior

An error is reported when indexing `PartialOnUndefined<T>` with type `"foo"` where generic type parameter `T` extends `{ foo: string; }`:
```
Type '"foo"' cannot be used to index type 'PartialOnUndefined<T>'. (2536)
```

### 🙂 Expected behavior

There should be no error; TS should be able to determine that `"foo"` can always index `PartialOnUndefined<T>`, since `T` extends `{ foo: string; }`.

### Additional information about the issue

Possibly related to https://github.com/microsoft/TypeScript/issues/36631

Workarounds:
- Add a `ts-ignore` comment to silence the error; the type alias can still be used without issue.
- Explicitly narrow `PartialOnUndefined<T>`:
```ts
type Generic<T extends { foo: string; }>
    = PartialOnUndefined<T> extends { foo: string; }
        ? PartialOnUndefined<T>["foo"]
        : never;
```

**Comment by RyanCavanaugh** [comment]
I don't think we'll ever have the kind of logic that would be needed to determine that `keyof PartialOnUndefined<T>` was exactly the same as `keyof T`. Inspecting the two index signatures here to figure out that they are exactly complementary is too niche of a situation.
